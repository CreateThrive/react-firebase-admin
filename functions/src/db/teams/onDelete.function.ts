import { database } from 'firebase-admin';
import { database as db } from 'firebase-functions';
import { BatchUpdate, IUserRtdb } from '../../types';

export default db.ref('teams/{teamId}').onDelete(async (snapshot, context) => {
  const { teamId } = context.params;
  let users;

  try {
    users = (await database().ref('users').once('value')).val();
  } catch (error) {
    console.log(error);
  }

  users = Object.entries(users).filter(([key, user]) => {
    const { teams } = user as IUserRtdb;

    return teams?.includes(teamId);
  });

  const batchUpdate: BatchUpdate = {};

  users.forEach(([key, user]) => {
    const { teams } = user as IUserRtdb;

    if (teams) {
      const newTeams = teams.filter((index: string) => index !== teamId);

      batchUpdate[`/${key}/teams`] = newTeams;
    }
  });

  return database().ref('users').update(batchUpdate);
});
