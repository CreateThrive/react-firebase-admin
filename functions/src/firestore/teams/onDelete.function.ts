import { firestore } from 'firebase-admin';
import { firestore as firestoreFunction } from 'firebase-functions';
import { IUserFirestore } from '../../types';

export default firestoreFunction
  .document('teams/{teamId}')
  .onDelete(async (snapshot, context) => {
    const { teamId } = context.params;

    const batch = firestore().batch();

    let snapshotsUsers;
    try {
      snapshotsUsers = await firestore()
        .collection('users')
        .where('teams', 'array-contains', teamId)
        .get();
    } catch (error) {
      console.log(error);
    }

    snapshotsUsers?.forEach((userSnapshot: any) => {
      const { id } = userSnapshot;
      const { teams } = userSnapshot.data() as IUserFirestore;

      const newTeams = teams.filter((index: string) => index !== teamId);

      const userRef = firestore().collection('users').doc(id);

      batch.update(userRef, { teams: newTeams });
    });

    return batch.commit();
  });
