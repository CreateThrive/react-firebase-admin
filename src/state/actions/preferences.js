import { createAction } from 'redux-act';

export const PREFERENCES_SET_LOCALE = createAction('PREFERENCES_SET_LOCALE');

export const setUserLocale = locale => dispatch => {
  return dispatch(PREFERENCES_SET_LOCALE({ locale }));
};
