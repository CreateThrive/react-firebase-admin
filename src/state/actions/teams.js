import { createAction } from 'redux-act';
import { toastr } from 'react-redux-toastr';
import { v4 as uuid } from 'uuid';

import { firebaseError, deleteLogo, getLogoUrl, uploadLogo } from 'utils';
import {
  fetchCollection,
  createDocument,
  deleteDocument,
  updateDocument,
  batchUpdateDocument,
} from '../api';
import { removeStoreUser, addStoreUser } from './users';

export const TEAMS_FETCH_TEAMS_INIT = createAction('TEAMS_FETCH_TEAMS_INIT');

export const TEAMS_FETCH_TEAMS_SUCCESS = createAction(
  'TEAMS_FETCH_TEAMS_SUCCESS'
);

export const TEAMS_FETCH_TEAMS_FAIL = createAction('TEAMS_FETCH_TEAMS_FAIL');

export const TEAMS_DELETE_TEAM_INIT = createAction('TEAMS_DELETE_TEAM_INIT');

export const TEAMS_DELETE_TEAM_SUCCESS = createAction(
  'TEAMS_DELETE_TEAM_SUCCESS'
);

export const TEAMS_DELETE_TEAM_FAIL = createAction('TEAMS_DELETE_TEAM_FAIL');

export const TEAMS_CREATE_TEAM_INIT = createAction('TEAMS_CREATE_TEAM_INIT');

export const TEAMS_CREATE_TEAM_SUCCESS = createAction(
  'TEAMS_CREATE_TEAM_SUCCESS'
);
export const TEAMS_CREATE_TEAM_FAIL = createAction('TEAMS_CREATE_TEAM_FAIL');

export const TEAMS_MODIFY_TEAM_INIT = createAction('TEAMS_MODIFY_TEAM_INIT');

export const TEAMS_MODIFY_TEAM_SUCCESS = createAction(
  'TEAMS_MODIFY_TEAM_SUCCESS'
);
export const TEAMS_MODIFY_TEAM_FAIL = createAction('TEAMS_MODIFY_TEAM_FAIL');

export const TEAMS_CLEAN_UP = createAction('TEAMS_CLEAN_UP');

export const TEAMS_ADD_USER_INIT = createAction('TEAMS_ADD_USER_INIT');

export const TEAMS_ADD_USER_SUCCESS = createAction('TEAMS_ADD_USER_SUCCESS');

export const TEAMS_ADD_USER_FAIL = createAction('TEAMS_ADD_USER_FAIL');

export const TEAMS_REMOVE_USER_INIT = createAction('TEAMS_REMOVE_USER_INIT');

export const TEAMS_REMOVE_USER_SUCCESS = createAction(
  'TEAMS_REMOVE_USER_SUCCESS'
);

export const TEAMS_REMOVE_USER_FAIL = createAction('TEAMS_REMOVE_USER_FAIL');

export const TEAMS_SELECTED_TEAM = createAction('TEAMS_SELECTED_TEAM');

export const TEAMS_FETCH_TEAM_USERS_INIT = createAction(
  'TEAMS_FETCH_TEAM_USERS_INIT'
);

export const TEAMS_FETCH_TEAM_USERS_SUCCESS = createAction(
  'TEAMS_FETCH_TEAM_USERS_SUCCESS'
);

export const TEAMS_FETCH_TEAM_USERS_FAIL = createAction(
  'TEAMS_FETCH_TEAM_USERS_FAIL'
);

export const fetchTeams = () => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_FETCH_TEAMS_INIT());
    const { locale } = getState().preferences;
    const { isAdmin, id } = getState().auth.userData;

    let teams;
    try {
      teams = await fetchCollection(
        'teams',
        isAdmin
          ? {
              filterBy: 'users',
              value: id,
              isArray: true,
            }
          : {}
      );
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_FETCH_TEAMS_FAIL({ error }));
    }

    return dispatch(
      TEAMS_FETCH_TEAMS_SUCCESS({
        teams: isAdmin
          ? teams
          : teams.filter((value) => value.users?.includes(id)),
      })
    );
  };
};

export const deleteTeam = (id) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_DELETE_TEAM_INIT());
    const { locale } = getState().preferences;
    const { logoUrl } = getState().teams.teamsList.find(
      (team) => team.id === id
    );

    const deleteLogoTask = logoUrl ? deleteLogo(logoUrl, 'teams') : null;

    const deleteTeamTask = deleteDocument('teams', id);

    try {
      await Promise.all([deleteLogoTask, deleteTeamTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        TEAMS_DELETE_TEAM_FAIL({
          error: errorMessage,
        })
      );
    }

    toastr.success('', 'The team was deleted.');
    return dispatch(TEAMS_DELETE_TEAM_SUCCESS({ id }));
  };
};

export const createTeam = ({ name, description, file }) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_CREATE_TEAM_INIT());
    const { locale } = getState().preferences;

    const uid = uuid();

    let uploadLogoTask = null;
    let logoUrl = null;
    if (file) {
      logoUrl = getLogoUrl(uid, file, 'teams');
      uploadLogoTask = uploadLogo(uid, file, 'teams');
    }

    const teamData = {
      name,
      description,
      logoUrl,
      createdAt: new Date().toString(),
      createdBy: getState().auth.userData.name,
    };

    const createTeamDbTask = createDocument('teams', uid, teamData);

    try {
      await Promise.all([uploadLogoTask, createTeamDbTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        TEAMS_CREATE_TEAM_FAIL({
          error: errorMessage,
        })
      );
    }

    toastr.success('', 'Team created successfully');
    return dispatch(
      TEAMS_CREATE_TEAM_SUCCESS({ team: { ...teamData, id: uid } })
    );
  };
};

export const modifyTeam = ({ name, description, file, id }) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_MODIFY_TEAM_INIT());
    const { locale } = getState().preferences;
    const { logoUrl } = getState().teams.teamsList.find(
      (team) => team.id === id
    );

    let deleteLogoTask;
    let uploadLogoTask;
    let newLogoUrl = null;
    if (file) {
      newLogoUrl = getLogoUrl(id, file, 'teams');
      deleteLogoTask = logoUrl && deleteLogo(logoUrl, 'teams');
      uploadLogoTask = uploadLogo(id, file, 'teams');
    }

    const teamData = {
      name,
      description,
      logoUrl: logoUrl || newLogoUrl,
    };
    const updateTeamDbTask = updateDocument('teams', id, teamData);

    try {
      await Promise.all([deleteLogoTask, uploadLogoTask, updateTeamDbTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        TEAMS_MODIFY_TEAM_FAIL({
          error: errorMessage,
        })
      );
    }

    toastr.success('', 'Team updated successfully');

    return dispatch(TEAMS_MODIFY_TEAM_SUCCESS({ team: teamData, id }));
  };
};

export const teamsCleanUp = () => (dispatch) => dispatch(TEAMS_CLEAN_UP());

export const addUser = (teamId, user) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_ADD_USER_INIT());
    const { locale } = getState().preferences;
    const { teamsList } = getState().teams;

    const userId = user.id;

    const teamUsers =
      teamsList.find((value) => value.id === teamId)?.users || [];
    const userTeams = user?.teams || [];

    const newUserTeams = [...userTeams, teamId];
    const newTeamUsers = [...teamUsers, userId];

    const batchCollections = {
      users: {
        documents: [
          {
            id: userId,
            field: 'teams',
            value: newUserTeams,
          },
        ],
      },

      teams: {
        documents: [
          {
            id: teamId,
            field: 'users',
            value: newTeamUsers,
          },
        ],
      },
    };

    try {
      await batchUpdateDocument(batchCollections);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_ADD_USER_FAIL({ error }));
    }

    dispatch(removeStoreUser(userId));

    return dispatch(
      TEAMS_ADD_USER_SUCCESS({
        teamId,
        users: newTeamUsers,
        user: { ...user, teams: newUserTeams },
      })
    );
  };
};

export const removeUser = (teamId, userId) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_REMOVE_USER_INIT());
    const { locale } = getState().preferences;
    const { teamsList, usersList } = getState().teams;

    const user = usersList.find((value) => value.id === userId);
    const teamUsers =
      teamsList.find((value) => value.id === teamId)?.users || [];
    const userTeams = user?.teams || [];

    const newUserTeams = userTeams.filter((value) => value !== teamId);
    const newTeamUsers = teamUsers.filter((value) => value !== userId);

    const bulkUpdate = {};
    bulkUpdate[`/users/${userId}/teams`] = newUserTeams;
    bulkUpdate[`/teams/${teamId}/users`] = newTeamUsers;

    try {
      await updateDocument('', '', bulkUpdate);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_REMOVE_USER_FAIL({ error }));
    }

    dispatch(addStoreUser({ ...user, teams: newUserTeams }));

    return dispatch(
      TEAMS_REMOVE_USER_SUCCESS({
        teamId,
        users: newTeamUsers,
        userId,
      })
    );
  };
};

export const fetchTeamUsers = (teamId) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_FETCH_TEAM_USERS_INIT());
    const { locale } = getState().preferences;

    let users;
    try {
      users = await fetchCollection('users', {
        filterBy: 'users',
        value: teamId,
        isArray: true,
      });
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_FETCH_TEAM_USERS_FAIL({ error }));
    }

    dispatch(TEAMS_SELECTED_TEAM({ teamId }));

    return dispatch(
      TEAMS_FETCH_TEAM_USERS_SUCCESS({
        users: users.filter((value) => value.teams?.includes(teamId)),
      })
    );
  };
};
