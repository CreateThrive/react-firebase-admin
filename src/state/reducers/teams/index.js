import { createReducer } from 'redux-act';

import {
  TEAMS_FETCH_TEAMS_INIT,
  TEAMS_FETCH_TEAMS_SUCCESS,
  TEAMS_FETCH_TEAMS_FAIL,
  TEAMS_DELETE_TEAM_INIT,
  TEAMS_DELETE_TEAM_SUCCESS,
  TEAMS_DELETE_TEAM_FAIL,
  TEAMS_CREATE_TEAM_INIT,
  TEAMS_CREATE_TEAM_SUCCESS,
  TEAMS_CREATE_TEAM_FAIL,
  TEAMS_MODIFY_TEAM_INIT,
  TEAMS_MODIFY_TEAM_SUCCESS,
  TEAMS_MODIFY_TEAM_FAIL,
  TEAMS_CLEAN_UP,
  TEAMS_ADD_USER_FAIL,
  TEAMS_ADD_USER_INIT,
  TEAMS_ADD_USER_SUCCESS,
  TEAMS_REMOVE_USER_FAIL,
  TEAMS_REMOVE_USER_INIT,
  TEAMS_REMOVE_USER_SUCCESS,
  TEAMS_SELECTED_TEAM,
  TEAMS_FETCH_TEAM_USERS_FAIL,
  TEAMS_FETCH_TEAM_USERS_INIT,
  TEAMS_FETCH_TEAM_USERS_SUCCESS,
} from 'state/actions/teams';

const initialState = {
  teamsList: [],
  usersList: [],
  loadingTeams: false,
  loadingUsers: false,
  error: null,
  success: false,
  deleted: false,
  teamId: 0,
};

export const teamsReducer = createReducer(
  {
    [TEAMS_FETCH_TEAMS_INIT]: () => ({
      ...initialState,
      loadingTeams: true,
    }),
    [TEAMS_FETCH_TEAMS_SUCCESS]: (state, payload) => ({
      ...state,
      teamsList: payload.teams,
      loadingTeams: false,
      error: null,
    }),
    [TEAMS_FETCH_TEAMS_FAIL]: (state, payload) => ({
      ...state,
      loadingTeams: false,
      error: payload.error,
    }),
    [TEAMS_DELETE_TEAM_INIT]: (state) => ({
      ...state,
      loadingTeams: true,
    }),
    [TEAMS_DELETE_TEAM_SUCCESS]: (state, payload) => ({
      ...state,
      teamsList: state.teamsList.filter((elem) => elem.id !== payload.id),
      loadingTeams: false,
      error: null,
      deleted: true,
    }),
    [TEAMS_DELETE_TEAM_FAIL]: (state, payload) => ({
      ...state,
      loadingTeams: false,
      error: payload.error,
    }),
    [TEAMS_CREATE_TEAM_INIT]: (state) => ({
      ...state,
      loadingTeams: true,
    }),
    [TEAMS_CREATE_TEAM_SUCCESS]: (state, payload) => ({
      ...state,
      teamsList: state.teamsList.concat(payload.team),
      loadingTeams: false,
      error: null,
      success: true,
    }),
    [TEAMS_CREATE_TEAM_FAIL]: (state, payload) => ({
      ...state,
      loadingTeams: false,
      error: payload.error,
    }),
    [TEAMS_MODIFY_TEAM_INIT]: (state) => ({
      ...state,
      loadingTeams: true,
    }),
    [TEAMS_MODIFY_TEAM_SUCCESS]: (state, payload) => ({
      ...state,
      teamsList: !state.teamsList
        ? []
        : state.teamsList.map((elem) => {
            if (elem.id === payload.id) {
              return {
                ...elem,
                name: payload.team.name,
                description: payload.team.description,
                logoUrl: payload.team.logoUrl || null,
              };
            }
            return elem;
          }),
      loadingTeams: false,
      error: null,
      success: true,
    }),
    [TEAMS_MODIFY_TEAM_FAIL]: (state, payload) => ({
      ...state,
      loadingTeams: false,
      error: payload.error,
    }),
    [TEAMS_CLEAN_UP]: (state) => ({
      ...state,
      loadingTeams: false,
      error: null,
      success: false,
      deleted: false,
    }),
    [TEAMS_ADD_USER_FAIL]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: payload.error,
    }),
    [TEAMS_ADD_USER_INIT]: (state) => ({
      ...state,
      loadingUsers: true,
    }),
    [TEAMS_ADD_USER_SUCCESS]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: null,
      success: true,
      teamsList: state.teamsList.map((value) => {
        if (value.id === payload.teamId) {
          return { ...value, users: payload.users };
        }
        return value;
      }),
      usersList: [...state.usersList, payload.user],
    }),
    [TEAMS_REMOVE_USER_FAIL]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: payload.error,
    }),
    [TEAMS_REMOVE_USER_INIT]: (state) => ({
      ...state,
      loadingUsers: true,
    }),
    [TEAMS_REMOVE_USER_SUCCESS]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: null,
      teamsList: state.teamsList.map((value) => {
        if (value.id === payload.teamId) {
          return { ...value, users: payload.users };
        }
        return value;
      }),
      usersList: state.usersList.filter((value) => value.id !== payload.userId),
      deleted: true,
    }),
    [TEAMS_SELECTED_TEAM]: (state, payload) => ({
      ...state,
      teamId: payload.teamId,
    }),
    [TEAMS_FETCH_TEAM_USERS_INIT]: (state) => ({
      ...state,
      loadingUsers: true,
    }),
    [TEAMS_FETCH_TEAM_USERS_SUCCESS]: (state, payload) => ({
      ...state,
      usersList: payload.users || [],
      loadingUsers: false,
      error: null,
    }),
    [TEAMS_FETCH_TEAM_USERS_FAIL]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: payload.error,
    }),
  },
  initialState
);
