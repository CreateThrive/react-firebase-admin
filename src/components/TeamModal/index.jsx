/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react';
import PropTypes from 'prop-types';
import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers';
import * as yup from 'yup';
import classNames from 'classnames';

import ErrorMessage from 'components/ErrorMessage';
import { useFormatMessage } from 'hooks';
import { useSelector, shallowEqual, useDispatch } from 'react-redux';
import { createTeam, modifyTeam } from 'state/actions/teams';

const schema = yup.object().shape({
  name: yup.string().required(),
  description: yup.string().notRequired(),
});

const TeamModal = ({ isActive, onCancel, isEditing }) => {
  const { teamsList, teamId, loading } = useSelector(
    (state) => ({
      teamsList: state.teams.teamsList,
      teamId: state.teams.teamId,
      loading: state.teams.loadingTeams,
    }),
    shallowEqual
  );

  const dispatch = useDispatch();

  const { register, handleSubmit, errors, watch } = useForm({
    defaultValues: isEditing
      ? teamsList.find((value) => value.id === teamId)
      : {},
    resolver: yupResolver(schema),
  });

  const onSubmitHandler = (value) => {
    const team = {
      ...value,
      file: value?.file[0] || null,
    };
    dispatch(isEditing ? modifyTeam(team) : createTeam(team));
  };

  const pickAnotherFileMessage = useFormatMessage('UserForm.pickAnotherFile');

  const pickFileMessage = useFormatMessage('UserForm.pickFile');

  const addTeamButtonMessage = useFormatMessage('TeamModal.addTeam');

  const updateTeamButtonMessage = useFormatMessage('TeamModal.updateTeam');

  const cancelButtonMessage = useFormatMessage('TeamModal.cancel');

  return (
    <div className={classNames(`modal`, { 'is-active': isActive })}>
      <div
        className="modal-background"
        onClick={!loading ? onCancel : undefined}
      />
      <div className="modal-card">
        <header className="modal-card-head">
          <span>{useFormatMessage('TeamModal.teamInformation')}</span>
        </header>
        <form onSubmit={handleSubmit(onSubmitHandler)}>
          <section className="modal-card-body">
            <div className="field is-horizontal">
              <div className="field-label is-normal">
                <label className="label">
                  {useFormatMessage('TeamModal.name')}
                </label>
              </div>
              <div className="field-body">
                <div className="field">
                  <div className="control">
                    <input
                      className="input"
                      type="text"
                      ref={register}
                      name="name"
                    />
                  </div>
                </div>
              </div>
            </div>
            {errors.name && (
              <div className="field is-horizontal">
                <div className="field-label is-normal" />
                <div className="field-body">
                  <ErrorMessage />
                </div>
              </div>
            )}
            <div className="field is-horizontal">
              <div className="field-label is-normal">
                <label className="label">
                  {useFormatMessage('TeamModal.description')}
                </label>
              </div>
              <div className="field-body">
                <div className="field">
                  <div className="control">
                    <input
                      className="input"
                      type="text"
                      ref={register}
                      name="description"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div className="field is-horizontal">
              <div className="field-label is-normal">
                <label className="label">
                  {useFormatMessage('UserForm.logo')}
                </label>
              </div>
              <div className="field-body">
                <div className="field">
                  <div className="file has-name">
                    <label className="file-label">
                      <input
                        className="file-input"
                        type="file"
                        name="file"
                        ref={register}
                        accept="image/*"
                      />
                      <span className="file-cta">
                        <span className="file-icon">
                          <i className="mdi mdi-upload" />
                        </span>
                        <span className="file-label">
                          {watch('file') && watch('file').file
                            ? pickAnotherFileMessage
                            : pickFileMessage}
                        </span>
                      </span>
                      <span className="file-name">
                        {watch('file') && watch('file')[0]?.name}
                      </span>
                    </label>
                  </div>
                </div>
              </div>
            </div>
          </section>
          <footer className="modal-card-foot">
            <button
              type="submit"
              className={classNames(`button is-success`, {
                'is-loading': loading,
              })}
            >
              {!isEditing ? addTeamButtonMessage : updateTeamButtonMessage}
            </button>
            <button
              type="button"
              className="button"
              onClick={onCancel}
              disabled={loading}
            >
              {cancelButtonMessage}
            </button>
          </footer>
        </form>
      </div>
    </div>
  );
};

TeamModal.propTypes = {
  isActive: PropTypes.bool.isRequired,
  onCancel: PropTypes.func.isRequired,
  isEditing: PropTypes.bool,
};

TeamModal.defaultProps = {
  isEditing: false,
};

export default TeamModal;
