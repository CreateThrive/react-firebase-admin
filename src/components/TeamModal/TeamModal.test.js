import { fireEvent } from '@testing-library/react';
import React from 'react';
import * as reactRedux from 'react-redux';
import '@testing-library/jest-dom';

import * as actions from 'state/actions/teams';
import TeamModal from '.';

describe('<TeamModal /> rendering', () => {
  const dispatchMock = jest.fn();
  let teamData;

  const state = {
    auth: { userData: { isAdmin: true } },
    teams: {
      success: false,
      deleted: false,
      loadingUsers: false,
      loadingTeams: false,
      teamsList: ['testTeam'],
      usersList: [{ name: 'testUser' }],
    },
    users: { data: ['testUser'] },
  };

  beforeEach(() => {
    jest
      .spyOn(reactRedux, 'useDispatch')
      .mockImplementation(() => dispatchMock);
    jest.spyOn(actions, 'modifyTeam').mockImplementation(jest.fn);
    jest.spyOn(actions, 'createTeam').mockImplementation(jest.fn);

    teamData = {
      id: 'test id',
      name: 'Test name',
      description: 'Test description',
      createdBy: 'Test User',
      logoUrl: 'some logo',
      users: [],
    };
  });

  it('should render without crashing', () => {
    const { component } = renderWithProviders(
      <TeamModal isActive onCancel={jest.fn()} />
    )({
      ...state,
    });

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should disable Cancel button when loading', () => {
    const onCancel = jest.fn();
    const { component } = renderWithProviders(
      <TeamModal isActive onCancel={onCancel} />
    )({
      ...state,
      teams: {
        ...state.teams,
        loadingTeams: true,
      },
    });

    expect(component.getByText('Cancel')).toBeDisabled();
  });

  it('should call onCancel when clicking Cancel button', () => {
    const onCancel = jest.fn();
    const { component } = renderWithProviders(
      <TeamModal isActive onCancel={onCancel} />
    )({
      ...state,
    });

    fireEvent.click(component.getByText('Cancel'));

    expect(onCancel).toBeCalledTimes(1);
  });

  it('should display the add button loading correctly', () => {
    const onCancel = jest.fn();
    const { component } = renderWithProviders(
      <TeamModal isActive onCancel={onCancel} />
    )({
      ...state,
      teams: {
        ...state.teams,
        loadingTeams: true,
      },
    });

    expect(component.getByText('Add Team')).toHaveClass('is-loading');
  });

  it('should dispatch createTeam action when creating a new user', async () => {
    const { component } = renderWithProviders(
      <TeamModal isActive onCancel={jest.fn()} />
    )({ ...state });

    fireEvent.click(component.getByText('Add Team'));

    await (() => expect(actions.createTeam).toBeCalledTimes(1));

    await (() => expect(actions.createTeam).toBeCalledWith(teamData));
  });

  it('should dispatch modifyTeam action when editing a user', async () => {
    const { component } = renderWithProviders(
      <TeamModal isActive onCancel={jest.fn()} isEditing />
    )({ ...state });

    fireEvent.click(component.getByText('Update Team'));

    await (() => expect(actions.modifyTeam).toBeCalledTimes(1));

    await (() => expect(actions.modifyTeam).toBeCalledWith(teamData));
  });
});
