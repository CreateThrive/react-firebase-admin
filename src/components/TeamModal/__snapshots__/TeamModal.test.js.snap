// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TeamModal /> rendering should render without crashing 1`] = `
<DocumentFragment>
  <div
    class="modal is-active"
  >
    <div
      class="modal-background"
    />
    <div
      class="modal-card"
    >
      <header
        class="modal-card-head"
      >
        <span>
          Team Information
        </span>
      </header>
      <form>
        <section
          class="modal-card-body"
        >
          <div
            class="field is-horizontal"
          >
            <div
              class="field-label is-normal"
            >
              <label
                class="label"
              >
                Name
              </label>
            </div>
            <div
              class="field-body"
            >
              <div
                class="field"
              >
                <div
                  class="control"
                >
                  <input
                    class="input"
                    name="name"
                    type="text"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="field is-horizontal"
          >
            <div
              class="field-label is-normal"
            >
              <label
                class="label"
              >
                Description
              </label>
            </div>
            <div
              class="field-body"
            >
              <div
                class="field"
              >
                <div
                  class="control"
                >
                  <input
                    class="input"
                    name="description"
                    type="text"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="field is-horizontal"
          >
            <div
              class="field-label is-normal"
            >
              <label
                class="label"
              >
                Logo
              </label>
            </div>
            <div
              class="field-body"
            >
              <div
                class="field"
              >
                <div
                  class="file has-name"
                >
                  <label
                    class="file-label"
                  >
                    <input
                      accept="image/*"
                      class="file-input"
                      name="file"
                      type="file"
                    />
                    <span
                      class="file-cta"
                    >
                      <span
                        class="file-icon"
                      >
                        <i
                          class="mdi mdi-upload"
                        />
                      </span>
                      <span
                        class="file-label"
                      >
                        Pick a file
                      </span>
                    </span>
                    <span
                      class="file-name"
                    />
                  </label>
                </div>
              </div>
            </div>
          </div>
        </section>
        <footer
          class="modal-card-foot"
        >
          <button
            class="button is-success"
            type="submit"
          >
            Add Team
          </button>
          <button
            class="button"
            type="button"
          >
            Cancel
          </button>
        </footer>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
